### PR TITLE
MAGN-9067 Clicking on class name of search results...

### PR DESCRIPTION
**Pull request by aosyatnik**
### Purpose

Link to YouTrack:
[MAGN-9067](http://adsk-oss.myjetbrains.com/youtrack/issue/MAGN-9067) Clicking on class name of search results brings the user to the expanded tree view of that class

When user moves mouse over class name cursor is changed into hand; text is highlighted and underlined.
When user clicks on text, search view is closed, search text box is cleared. Selected class is opened and brought into view.

### Declarations

- [x] The code base is in a better state after this PR
- [ ] Is documented according to the [standards](https://github.com/DynamoDS/Dynamo/wiki/Coding-Standards)
- [x] The level of testing this PR includes is appropriate
- [ ] User facing strings, if any, are extracted into `*.resx` files
- [ ] All tests pass using the self-service CI.

### Reviewers

@ramramps 

### FYIs

@Racel

### CODE REVIEW

File: src/DynamoCoreWpf/ViewModels/Search/SearchViewModel.cs [hunk 1]
```diff
@@ -1089,6 +1089,37 @@ internal void SelectAllCategories(object parameter)
             }
         }
 
+        internal void OpenSelectedClass(string className)
```
Comment (ramramps) [inline src/DynamoCoreWpf/ViewModels/Search/SearchViewModel.cs]: Better to write comment about this class behavior

File: src/DynamoCoreWpf/ViewModels/Search/SearchViewModel.cs [hunk 2]
```diff
@@ -1089,6 +1089,37 @@ internal void SelectAllCategories(object parameter)
...
+        {
+            // Clear search text.
+            SearchText = String.Empty;
+            var categoryNames = className.Split(Configurations.CategoryDelimiterString.ToCharArray());
```
Comment (ramramps) [inline src/DynamoCoreWpf/ViewModels/Search/SearchViewModel.cs]: what is ClassName is empty ?

File: src/DynamoCoreWpf/Views/Search/LibrarySearchView.xaml.cs [hunk 3]
```diff
@@ -140,5 +140,35 @@ private void OnButtonPreviewMouseMove(object sender, MouseEventArgs e)
...
+            var searchElementVM = senderButton.DataContext as NodeSearchElementViewModel;
+            if (searchElementVM == null)
+            {
+                return;
+            }
+
+            int lastIndex =
+                searchElementVM.FullName.LastIndexOf(Configuration.Configurations.CategoryDelimiterString,
+                    StringComparison.Ordinal);
+            string selectedClassWithCategory;
```
Comment (ramramps) [inline src/DynamoCoreWpf/Views/Search/LibrarySearchView.xaml.cs]: move the initialization to the beginning of this method. say selectedClassWithCategory = string.empty

**Comment by ramramps** [comment]
Clicking a class name should bring the appropriate class into view. Ex : AdjacentEdges in vertex.  Clicking on vertex should  bring the expanded tree to the view. I suggest to adjust the scroll position with reference to that class.

**Comment by aosyatnik** [comment]
Hi, @ramramps . Yes, I have already thought about it, that's why I created `OnTreeViewItemExpanded`. Every time, when item is expanded it's brought into view. 
E.g.
image

**Comment by ramramps** [comment]
Here is the scenario 
 case 1 :Type AdjacentEdges. Click Vertex. You can see vertex expanded correctly. 
 case 2 : Type AdjacentFaces Click Vertex  You have to scroll to see the vertex category - Incorrect.
 
 case 3 : Now you collapse all the subcategories and categories. Again type AdjacentFaces. Click vertex. you can see the vertex expanded correctly. 

So the case 2 is incorrect.  Please check.

**Comment by aosyatnik** [comment]
Hi, @ramramps , I didn't think about this case. Thank you.

 AdjacentFaces is not scrolled for 2nd time, because it wasn't expanded for 2nd time. In first step we expanded its' category and since that time nothing has changed. `RaisePropertyChanged` wasn't called, because `IsExpaned` wasn't changed.

I think appropriate solution is first to collapse all categories and subcategories and then start to expand something. I created one more method for collapsing: `CollapseRecursively` and test for it.

File: src/DynamoCoreWpf/ViewModels/Search/SearchViewModel.cs [hunk 4]
```diff
@@ -1089,6 +1089,68 @@ internal void SelectAllCategories(object parameter)
             }
         }
 
+        /// <summary>
```
Comment (ramramps) [inline src/DynamoCoreWpf/ViewModels/Search/SearchViewModel.cs]: I understand this. But is there a way to check the expanded Subcategory and then the expanded category , and make it to false ?

Comment (aosyatnik) [inline src/DynamoCoreWpf/ViewModels/Search/SearchViewModel.cs]: @ramramps , Do you mean to start from the deepest layer, i.e.from the last subcategory? But what is the reason? In any case we have to set false for all of them and order doesn't matter. Or it does?

Comment (ramramps) [inline src/DynamoCoreWpf/ViewModels/Search/SearchViewModel.cs]: @aosyatnik - No. Why cannot we say false to only those categories where Isexpanded is set to true? 
Why we should traverse all the categories here?

Comment (aosyatnik) [inline src/DynamoCoreWpf/ViewModels/Search/SearchViewModel.cs]: @ramramps 
In any case, if you want to find something, then you need to search for something. If we want to set false for all expanded items, we must search for these expanded items. And in this case we will search among all items just  for expanded items. So, in any case we need to traverse all categories.
But! Another question: what is faster assignment or an if-statement?
http://stackoverflow.com/questions/4043455/which-costs-more-while-looping-assignment-or-an-if-statement

I decided to check :)
* Without if
search for point:
image
image
image
image

* With if
search for point:
image
image
image
image

I can't say, that I see big lose of time. So, I believe we can leave without if.

P.S. that was really interesting research :D

Comment (ramramps) [inline src/DynamoCoreWpf/ViewModels/Search/SearchViewModel.cs]: Interesting. Thanks for it. 
But my question here, in CollapseRecursiverly, you have all the categories.
why can't you get categories.where(x => x.IsExpanded = true); This gives the category that is expanded.
Now for that category get the subcategory and say IsExpanded = false, and then set IsExpanded = false on that category.

I am sorry that I still don't understand the need for searching all the categories here.

Comment (aosyatnik) [inline src/DynamoCoreWpf/ViewModels/Search/SearchViewModel.cs]: @ramramps , that's what I meant to say... "If we want to set false for all expanded items, we must search for these expanded items".
All in all, LINQ still uses loops inside. But to be sure, I'll check this also... I don't know what is faster foreach loop or LINQ.

Comment (ramramps) [inline src/DynamoCoreWpf/ViewModels/Search/SearchViewModel.cs]: LINQ is much faster here.

Comment (aosyatnik) [inline src/DynamoCoreWpf/ViewModels/Search/SearchViewModel.cs]: Hi, @ramramps , I can't say, that LINQ is much faster:
image
image
image
image
image

But it definitely looks better :) So, let's leave LINQ here.

File: src/DynamoCoreWpf/ViewModels/Search/SearchViewModel.cs [hunk 5]
```diff
@@ -1089,6 +1089,61 @@ internal void SelectAllCategories(object parameter)
...
+            while (categories != null)
+            {
+                var category = categories.FirstOrDefault(cat => cat.IsExpanded);
+
+                if (category == null) break;
+                category.IsExpanded = false;
+
+                categories = category.SubCategories;
+            }
+        }
```
Comment (ramramps) [inline src/DynamoCoreWpf/ViewModels/Search/SearchViewModel.cs]: Much better version.

**Pull request merged**